qubes-gui-client/src/buffer.rs

```diff
@@ -0,0 +1,389 @@
+/*
+ * The Qubes OS Project, https://www.qubes-os.org
+ *
+ * Copyright (C) 2021  Demi Marie Obenour  <demi@invisiblethingslab.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ */
+//! A wrapper around vchans that provides a write buffer.  Used to prevent
+//! deadlocks.
+
+use qubes_castable::Castable as _;
+use qubes_gui::Header;
+use std::collections::VecDeque;
+use std::io::{self, Error, ErrorKind};
+use std::mem::size_of;
+use std::ops::Range;
+
+#[derive(Debug)]
+enum ReadState {
+    ReadingHeader,
+    ReadingBody(Header, usize),
+    Discard(usize),
+    Error,
+}
+
+// Trait for a vchan, for unit-testing
+pub(crate) trait VchanMock {
+    fn buffer_space(&self) -> usize;
+    fn recv(&mut self, buf: &mut [u8]) -> io::Result<usize>;
+    fn send(&mut self, buf: &[u8]) -> io::Result<usize>;
+    fn wait(&self);
+    fn data_ready(&self) -> usize;
+}
+
+impl VchanMock for vchan::Vchan {
+    fn buffer_space(&self) -> usize {
+        vchan::Vchan::buffer_space(self)
+    }
+    fn recv(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        vchan::Vchan::recv(self, buf)
+    }
+    fn send(&mut self, buf: &[u8]) -> io::Result<usize> {
+        vchan::Vchan::send(self, buf)
+    }
+    fn wait(&self) {
+        vchan::Vchan::wait(self)
+    }
+    fn data_ready(&self) -> usize {
+        vchan::Vchan::data_ready(self)
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct Vchan<T: VchanMock> {
```

Comment:
The naming conflict between `buffer::Vchan` and `vchan::Vchan` might get confusing. Maybe `BufferedVchan`? `FramingVchan`? (i.e. a vchan that frames)

Reply:
Also perhaps: `RawMessageStream`?